### PR TITLE
121 - Code style adjustments and refactoring

### PR DESCRIPTION
**Pull request by nmaarnio**
Some code style adjustments and refactoring was made following the discussion in [Improving contribution and style guide issue](https://github.com/GispoCoding/eis_toolkit/issues/121) with some additional changes. 

Changes in this PR:
- Remove _prediction_methods_ and _model_training_ folders
- Add beartype to public functions for automatic argument type checking
- Remove type annotations from function parameters in docstrings (MkDocs can fetch the types from function signature)
- Remove return variable names from docstrings
- Add a missing .md file
- Add comments to exceptions
- Remove, add and order some custom exceptions
- Remove non-copy option from some utility functions
- Refactor and improve some functions, including:
  - Logic
  - Variable names
  - Documentation
  - Testing

### CODE REVIEW

**Comment by nmaarnio** [comment]
Regarding pull requests and repository history, do you think we should start squashing commits when merging to the master branch? I have not needed this feature before, but it seems like a good option to keep the repository history cleaner.

**Comment by nialov** [comment]
Squashing is quite an advanced `git` concept so I do not recommend knowing how to rebase `git` commits during a pull request to be a requirement for contributors. It is easy (-ish) when the commits you want to squash are in serial order and difficult when they are not. I suppose maintainters can edit branches that are committed to the main repository (I have used `nialov/eis_toolkit` out of habit) and can do squashing if needed. I would apply it selectively when you believe there is need.

I would also not by default squash a pull request to a single commit as the commit messages can be useful in future debugging.

**Comment by nmaarnio** [comment]
@Mtk112 promised to review this PR, but if you would also have time @nialov to take a quick look it would be appreciated since you have been part of the style & utility functions discussions. I attempted to address everything we discussed (and a few other things), except I did not change our checks to raise exceptions yet.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -212,3 +217,55 @@ poetry export --without-hashes --format=requirements.txt > requirements.txt
 and committing the new version of the particular file into your feature
 branch. Please note that this file is only used for GitHub workflows, otherwise
 we utilize poetry for dependency  handling.
+
+
+## Recent changes
+Some changes have been made to the style guide:
```
Comment (nialov) [inline CONTRIBUTING.md:223]: I think all the conventions applied in this pull request could be mentioned, i.e.

```suggestion
Some changes have been made to the style guide:
- Use `numbers.Number` as the type when both floats and integers are accepted by functions:
```python
from numbers import Number

def func(int_or_float: Number):
    ...
```
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -212,3 +217,55 @@ poetry export --without-hashes --format=requirements.txt > requirements.txt
...
+```python
+raise InvalidParameterValueException(f"Window size is too small: {height}, {width}.")
+```
+- Use beartype's decorator for automatic function argument type checking:
+```python
+from beartype import beartype
+
+@beartype
+def my_function(parameter_1: float, parameter_2: bool):
+```
```
Comment (nialov) [inline CONTRIBUTING.md:234]:
```suggestion
- Use beartype's decorator for automatic function argument type checking and import types from `beartype.typing` if a warning is raised by beartype on imports from `typing`:
```python
from beartype import beartype
from beartype.typing import Sequence

@beartype
def my_function(parameter_1: float, parameter_2: bool, parameter_seq: Sequence):
```
```

File: tests/checks/check_raster_grids_test.py [hunk 3]
```diff
@@ -5,16 +5,12 @@
...
-snap_raster = rasterio.open(parent_dir.joinpath("data/remote/snapping/snap_raster.tif"))
-small_raster = rasterio.open(parent_dir.joinpath("data/remote/small_raster.tif"))
-small_raster_epsg4326 = rasterio.open(parent_dir.joinpath("data/remote/small_raster_EPSG4326.tif"))
-snap_raster_smaller_cells = rasterio.open(
-    parent_dir.joinpath("data/remote/snapping/snap_test_raster_smaller_cells.tif")
-)
-snap_raster_top_right = rasterio.open(parent_dir.joinpath("data/remote/snapping/snap_test_raster_right_top.tif"))
-snap_raster_bottom_right = rasterio.open(parent_dir.joinpath("data/remote/snapping/snap_test_raster_right_bottom.tif"))
-clipped_snap_raster = rasterio.open(parent_dir.joinpath("data/remote/snapping/clipped_snap_raster.tif"))
+test_dir = Path(__file__).parent.parent
```
Comment (nialov) [inline tests/checks/check_raster_grids_test.py:8]: My (personal) convention is to uppercase any module symbols to differentiate them inside functions from local symbols. E.g. `test_dir` -> `TEST_DIR`. Maybe worth considering?

**Review by nialov — CHANGES_REQUESTED** [review]

**Comment by nialov** [comment]
Looks great overall and the naming conventions make things much clearer. I "requested changes" but only the mentioning of `numbers.Number` and `beartype.typing` are actually important. Rest are minor subjective stuff

**Review by Mtk112 — CHANGES_REQUESTED** [review]
Looks good, only managed to find one additional typo to fix.

**Comment by nmaarnio** [comment]
Thanks for the reviews, I fixed all that you pointed out. I didn't rename any module symbols yet, but let's keep that in mind.

**Review by Mtk112 — APPROVED** [review]
Looks good!

**Pull request merged**